### PR TITLE
[DPE-2778] Shard removal tests

### PR DESCRIPTION
**Pull request by MiaAltieri**
## Issue
No tests for shard removal

## Solution
Add tests for various removal scenarios

## Extras
1. These tests highlighted an issue with slow drainage. A large TIMEOUT period has been added to the tests and an issue (#282 ) has been added to the repo 
2. Wait for relation_changed to be given both key and ops password, as these seem to not always be given from the config server on relation-joined. [discussed here](https://chat.charmhub.io/charmhub/pl/prmxufjyfbnt9qrzdrxoxkqz3e)

### CODE REVIEW

File: tests/integration/sharding_tests/test_sharding.py [hunk 1]
```diff
@@ -21,23 +31,27 @@ async def test_build_and_deploy(ops_test: OpsTest) -> None:
     my_charm = await ops_test.build_charm(".")
     await ops_test.model.deploy(
         my_charm,
-        num_units=2,
+        num_units=1,
```
Comment (MiaAltieri) [inline tests/integration/sharding_tests/test_sharding.py]: change back to `2`

File: tests/integration/sharding_tests/test_sharding.py [hunk 2]
```diff
@@ -21,23 +31,27 @@ async def test_build_and_deploy(ops_test: OpsTest) -> None:
...
         config={"role": "config-server"},
         application_name=CONFIG_SERVER_APP_NAME,
     )
     await ops_test.model.deploy(
-        my_charm, num_units=2, config={"role": "shard"}, application_name=SHARD_ONE_APP_NAME
+        my_charm, num_units=1, config={"role": "shard"}, application_name=SHARD_ONE_APP_NAME
```
Comment (MiaAltieri) [inline tests/integration/sharding_tests/test_sharding.py]: change back to `2`

File: tests/integration/sharding_tests/test_sharding.py [hunk 3]
```diff
@@ -21,23 +31,27 @@ async def test_build_and_deploy(ops_test: OpsTest) -> None:
...
     )
     await ops_test.model.deploy(
-        my_charm, num_units=2, config={"role": "shard"}, application_name=SHARD_TWO_APP_NAME
+        my_charm, num_units=1, config={"role": "shard"}, application_name=SHARD_TWO_APP_NAME
+    )
+    await ops_test.model.deploy(
+        my_charm, num_units=1, config={"role": "shard"}, application_name=SHARD_THREE_APP_NAME
```
Comment (MiaAltieri) [inline tests/integration/sharding_tests/test_sharding.py]: change back to `2`

File: lib/charms/mongodb/v1/shards_interface.py [hunk 4]
```diff
@@ -252,11 +253,19 @@ def remove_shards(self, departed_shard_id):
                     self.charm.unit.status = MaintenanceStatus(f"Draining shard {shard}")
                     logger.info("Attempting to removing shard: %s", shard)
                     mongo.remove_shard(shard)
+                except NotReadyError:
+                    logger.info("Unable to remove shard: %s another shard is draining", shard)
+                    # to guarantee that shard that the currently draining shard, gets re-processed,
+                    # do not raise immediately, instead at the end of removal processing.
+                    retry_removal = True
```
Comment (delgod) [inline lib/charms/mongodb/v1/shards_interface.py:258]: just Q: why not
```suggestion
                    raise ShardNotInClusterError
```

**Review by delgod — DISMISSED** [review]

File: lib/charms/mongodb/v1/shards_interface.py (hunk 4 above)
Comment (MiaAltieri) [reply lib/charms/mongodb/v1/shards_interface.py:258]: Does the comment above make sense?
```
# to guarantee that shard that the currently draining shard, gets re-processed,
# do not raise immediately, instead at the end of removal processing.
```

**Review by dmitry-ratushnyy — DISMISSED** [review]

**Review by delgod — APPROVED** [review]

**Review by dmitry-ratushnyy — APPROVED** [review]

**Pull request merged**